src/SFML/Audio/AudioDevice.cpp

```diff
@@ -221,6 +177,144 @@ ma_engine* AudioDevice::getEngine()
 }
 
 
+////////////////////////////////////////////////////////////
+bool AudioDevice::reinitialize()
+{
+    auto* instance = getInstance();
+
+    // We don't have to do anything if an instance doesn't exist yet
+    if (!instance)
+        return true;
+
+    const std::lock_guard lock(instance->m_resourcesMutex);
+
+    // Deinitialize all audio resources
+    for (const auto& entry : instance->m_resources)
+        entry.deinitializeFunc(entry.resource);
+
+    // Destroy the old engine
+    if (instance->m_engine)
+        ma_engine_uninit(&*instance->m_engine);
```

Comment:
This won't work. `std::unique_ptr` is not meant to store full objects like `std::optional` is.

Reply:
What do you mean? The unique_ptr will store a pointer to that full object. We’re changing allocation strategies but resource being managed can be the same.

Reply:
miniaudio doesn't allocate the objects itself, we would still have to `make_unique` them ourselves before passing them on to be initialized, and in the case initialization fails we would still have to immediately deallocate the uninitialized objects _without calling the `uninit` functions_. Besides the fact that having to dynamically allocate objects that are perfectly fine with automatic storage duration is just pure wasteful, this really isn't the place to abuse `std::unique_ptr` just for its deleter functionality. If we want RAII semantics then we should go the whole way and write proper wrappers instead of browsing through the standard library looking for constructs we can abuse just to save a few lines of code.